pallets/order-book/src/lib.rs

```diff
@@ -571,87 +551,111 @@ pub mod pallet {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			MarketFeederId::<T>::put(feeder_id);
+
+			Ok(())
 		}
 	}
 
 	impl<T: Config> Pallet<T> {
 		pub fn fulfill_order_with_amount(
 			order: OrderOf<T>,
-			buy_amount: T::Balance,
-			account_id: T::AccountId,
+			amount_out: T::Balance,
+			fulfilling_account: T::AccountId,
 		) -> DispatchResult {
 			ensure!(
-				buy_amount >= order.min_fulfillment_amount,
-				Error::<T>::InsufficientOrderSize,
+				amount_out >= order.min_fulfillment_amount_out,
+				Error::<T>::BelowMinFulfillmentAmount,
 			);
 
-			ensure!(
-				T::TradeableAsset::can_hold(order.asset_in_id, &(), &account_id, buy_amount),
-				Error::<T>::InsufficientAssetFunds,
-			);
+			let ratio = match order.ratio {
+				OrderRatio::Market => Self::market_ratio(order.asset_out_id, order.asset_in_id)?,
+				OrderRatio::Custom(ratio) => ratio,
+			};
 
-			let sell_amount = Self::convert_with_ratio(
-				order.asset_in_id,
-				order.asset_out_id,
-				order.max_sell_rate,
-				buy_amount,
-			)?;
-			let remaining_buy_amount = order
-				.buy_amount
-				.checked_sub(&buy_amount)
-				.ok_or(Error::<T>::BuyAmountTooLarge)?;
-			let partial_fulfillment = !remaining_buy_amount.is_zero();
+			let amount_in =
+				Self::convert_with_ratio(order.asset_out_id, order.asset_in_id, ratio, amount_out)?;
 
+			let remaining_amount_out = order
+				.amount_out
+				.checked_sub(&amount_out)
+				.ok_or(Error::<T>::FulfillAmountTooLarge)?;
+
+			let partial_fulfillment = !remaining_amount_out.is_zero();
 			if partial_fulfillment {
-				Self::update_order_with_fulfillment(
-					order.placing_account.clone(),
-					order.order_id,
-					remaining_buy_amount,
-					order.max_sell_rate,
-					remaining_buy_amount.min(order.min_fulfillment_amount),
-				)?;
-			} else {
-				T::TradeableAsset::release(
-					order.asset_out_id,
-					&(),
+				let mut updated_order = order.clone();
+				updated_order.amount_out = remaining_amount_out;
+				updated_order.amount_in = order.amount_in.ensure_add(amount_in)?;
+				updated_order.min_fulfillment_amount_out =
+					remaining_amount_out.min(order.min_fulfillment_amount_out);
+
+				<Orders<T>>::insert(updated_order.order_id, updated_order.clone());
+				<UserOrders<T>>::insert(
 					&order.placing_account,
-					sell_amount,
-					Precision::Exact,
-				)?;
-
+					updated_order.order_id,
+					updated_order,
+				);
+			} else {
 				Self::remove_order(order.order_id)?;
 			}
 
+			T::TradeableAsset::release(
+				order.asset_out_id,
+				&(),
+				&order.placing_account,
+				amount_out,
+				Precision::Exact,
+			)?;
+
+			if T::TradeableAsset::balance(order.asset_out_id, &order.placing_account) < amount_out {
+				Err(DispatchError::Token(TokenError::FundsUnavailable))?
+			}
+
+			if T::TradeableAsset::balance(order.asset_in_id, &fulfilling_account) < amount_in {
+				Err(DispatchError::Token(TokenError::FundsUnavailable))?
+			}
```

Comment:
This is sadly needed to protect against `placing_account` == `fulfilling_account`.

Reply:
Would there be a possible exploit here?

Reply:
Yes if we do not add this check.

The `placing_account` could fulfill their own order even if they have no funds.